### ISSUE TITLE
Document GitHub branching convention

### ISSUE THREAD

**Issue by kellijohnson-NOAA** [NOAA-FIMS/collaborative_workflow#6]
Choose and document a GitHub branching convention for the development of FIMS.
<a target="_blank" href="https://calendar.google.com/event?action=TEMPLATE&amp;tmeid=MnRjdThnbTVpcGZjZWp1Mm9lbmkyNWQ1cDUga2VsbGkuam9obnNvbkBub2FhLmdvdg&amp;tmsrc=kelli.johnson%40noaa.gov"><img border="0" src="https://www.google.com/calendar/images/ext/gc_button1_en.gif"></a>

**Comment by kellijohnson-NOAA** [comment]
## Branching strategy

There are several branching strategies available that will work within the Git environment, as well as other version control systems, but it is important to find a strategy that works well for both current and future contributors. Strategies provide guidance for how, when, and why branches are created and named, which also ties into necessary guidance surrounding issue tracking. **The trunk-based, also known as GitHub Flow, strategy was chosen for FIMS to make tasks easier without compromising quality.** This branching strategy facilitates knowledge of steps that must be taken prior to, during, and after making changes to the code; but, all who are interested in the code will be able to read it at any time.

There are perhaps another half a dozen branching strategies that are widely used and have worked well for other projects, including using a single branch, "feature branching", GitFlow, GitLab Flow, and OneFlow. Use of a single branch does not allow for protection of the published code base and GitFlow has two main branches, three supporting branches, and many feature branches with potentially long lags between merges. These extreme options are not ideal for this project, where trunk-based development captures the following needs while not begin overly complicated:
1. short-lived branches to minimize stale code and merge conflicts,
2. fast release times, especially for bug fixes,
3. ability to release bug fix without new features.
Additionally, trunk-based development is operable with the current skill set contained within the developers. Many references about this branching strategy exist on the web and in print.

### GitHub issue tracker

Use of the GitHub issue tracker is key to keeping everyone informed and prioritizing key tasks. Issues should be filed and tagged prior to any code changes whether the change pertains to a bug or the development of a feature. At a minimum, all issues will be labeled with a future version number. Bugs with immediate fixes will be assigned to the current version number augmented for a hot fix and development will be based on code in the trunk. All other issues will be assigned to a future version and development will be based on version branches. That is, changes to the code for version 3.3 cannot start until there is a branch for version 3.3. This will minimize stale code and large merge conflicts.

### Branch life

* short lived, less than 1 week?
* fully tested prior to merging
* reviewed
* tagged with feature labels

### Outside collaborators

Outside collaborators that are not part of the FIMS project will be required to fork the repository, make changes, and submit a pull request. Forks are discouraged for every-day development because it becomes difficult to keep track of all of the forks. Thus, it will be important for those working on forks to be active in the issue tracker in the main repository prior to working on their fork --- just like any member of the organization would do if they were working within the organization. Knowledge of future projects, ideas, concerns, etc. should always be documented in an issue before the code is altered.

Pull requests from forks will be reviewed the same as a pull request submitted from a branch. Users will need to conform to the same standards and all contributions must pass the standard tests as well as have tests that check the new feature.

**Comment by kellijohnson-NOAA** [comment]
### Tasks
- [ ] decide and reference a resource for trunk-based development where people can read about the workflow, e.g.,
  * Definition  
    * https://trunkbaseddevelopment.com/
  * Comparisons of methods
    * https://gitential.com/git-branching-strategies-for-your-team-how-to-choose-the-best/
    * https://www.atlassian.com/continuous-delivery/continuous-integration/trunk-based-development
- [ ] provide more detail
